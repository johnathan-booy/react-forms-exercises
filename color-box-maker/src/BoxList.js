import React, { useState } from "react";
import { v4 as uuid } from "uuid";
import Box from "./Box";
import "./BoxList.css";
import NewBoxForm from "./NewBoxForm";

const BoxList = () => {
	const INITIAL_STATE = [
		{ id: uuid(), color: "Aquamarine", width: 100, height: 100 },
		{ id: uuid(), color: "BlanchedAlmond", width: 120, height: 120 },
	];
	const [boxes, setBoxes] = useState(INITIAL_STATE);
	const addBox = (newBox) => {
		setBoxes((boxes) => [...boxes, { id: uuid(), ...newBox }]);
	};
	const removeBox = (deleteBox) => {
		setBoxes((boxes) => boxes.filter((box) => box.id !== deleteBox.id));
	};
	return (
		<div className="BoxList">
			<h3>Color Boxes</h3>
			<NewBoxForm addBox={addBox} />
			<div className="BoxList-boxes">
				{boxes.map(({ id, color, width, height }) => (
					<Box
						id={id}
						key={id}
						color={color}
						width={width}
						height={height}
						removeBox={removeBox}
					/>
				))}
			</div>
		</div>
	);
};

export default BoxList;
