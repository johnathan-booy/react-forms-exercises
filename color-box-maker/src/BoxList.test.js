import { fireEvent, getByLabelText, render } from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", () => {
	render(<BoxList />);
});

it("matches snapshot", () => {
	const { asFragment } = render(<BoxList />);
	expect(asFragment()).toMatchSnapshot();
});

it("renders NewBoxForm", () => {
	const { getByLabelText, getByText } = render(<BoxList />);

	expect(getByLabelText("Color")).toBeInTheDocument();
	expect(getByLabelText("Width")).toBeInTheDocument();
	expect(getByLabelText("Height")).toBeInTheDocument();
	expect(getByText("Add Box")).toBeInTheDocument();
});

it("renders default boxes", () => {
	const { getAllByRole } = render(<BoxList />);
	const box = getAllByRole("cell");
	expect(box).toHaveLength(2);
});

it("can add a new box", () => {
	const { getAllByRole, getByLabelText, getByText } = render(<BoxList />);

	// Should initialize with two boxes
	let boxes = getAllByRole("cell");
	expect(boxes).toHaveLength(2);

	// Create a new box with the form
	const colorInput = getByLabelText("Color");
	const addButton = getByText("Add Box");
	fireEvent.change(colorInput, "green");
	fireEvent.click(addButton);

	// Should have three boxes
	boxes = getAllByRole("cell");
	expect(boxes).toHaveLength(3);
});

it("can remove a box", () => {
	const { queryAllByRole, getAllByText } = render(<BoxList />);

	// Should initialize with two boxes
	let boxes = queryAllByRole("cell");
	expect(boxes).toHaveLength(2);

	// Create a new box with the form
	const removeButtons = getAllByText("X");
	removeButtons.map((btn) => fireEvent.click(btn));

	// Should not have any boxes
	boxes = queryAllByRole("cell");
	expect(boxes).toHaveLength(0);
});
