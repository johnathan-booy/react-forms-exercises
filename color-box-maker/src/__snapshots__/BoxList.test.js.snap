// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <h3>
      Color Boxes
    </h3>
    <form
      class="NewBoxForm"
    >
      <div
        class="NewBoxForm-group"
      >
        <label
          for="color"
        >
          Color
        </label>
        <input
          id="color"
          name="color"
          type="text"
          value=""
        />
      </div>
      <div
        class="NewBoxForm-group"
      >
        <label
          for="width"
        >
          Width
        </label>
        <input
          id="width"
          name="width"
          type="number"
          value="100"
        />
      </div>
      <div
        class="NewBoxForm-group"
      >
        <label
          for="height"
        >
          Height
        </label>
        <input
          id="height"
          name="height"
          type="number"
          value="100"
        />
      </div>
      <div
        class="NewBoxForm-group"
      >
        <button>
          Add Box
        </button>
      </div>
    </form>
    <div
      class="BoxList-boxes"
    >
      <div
        class="Box-group"
        role="row"
      >
        <div
          class="Box"
          role="cell"
          style="background-color: Aquamarine; width: 100px; height: 100px;"
        />
        <button
          class="Box-remove"
        >
          X
        </button>
      </div>
      <div
        class="Box-group"
        role="row"
      >
        <div
          class="Box"
          role="cell"
          style="background-color: BlanchedAlmond; width: 120px; height: 120px;"
        />
        <button
          class="Box-remove"
        >
          X
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
